backend/common/utils/spatial.py

```diff
@@ -0,0 +1,158 @@
+        self.bucket_name = f"spatial-deepzoom-{self.env}"
+        self.asset_directory = "spatial-deep-zoom"
+
+    def _calculate_aspect_ratio_crop(self, image_size):
+        """
+        Calculate the aspect ratio crop coordinates for an image.
+
+        Args:
+            image_size (int, int): The size of the image (width, height).
+
+        Returns:
+            tuple: The crop coordinates (left, upper, right, lower).
+        """
+        width, height = image_size
+        new_dimension = min(width, height)
+        left = (width - new_dimension) / 2
+        upper = (height - new_dimension) / 2
+        right = (width + new_dimension) / 2
+        lower = (height + new_dimension) / 2
+        return tuple(map(int, (left, upper, right, lower)))
+
+    def _prepare_image(self, content):
+        """
+        Prepare the image array for processing.
+
+        Args:
+            content (dict): The content dictionary containing the image array.
+
+        Returns:
+            np.ndarray: The prepared image array.
+        """
+        resolution = "fullres" if "fullres" in content["images"] else "hires"
+        image_array = content["images"][resolution]
+        image_array_uint8 = np.uint8(image_array * 255 if resolution == "hires" else image_array)
+        return image_array_uint8
+
+    def _process_and_flip_image(self, image_array_uint8):
+        """
+        Process and flip the image.
+
+        Args:
+            image_array_uint8 (np.ndarray): The image array.
+
+        Returns:
+            np.ndarray: The processed and flipped image array.
+        """
+        Image.MAX_IMAGE_PIXELS = None  # Disable the image size limit
+        try:
+            with Image.fromarray(image_array_uint8) as img:
+                # Convert image to RGB mode if it's not already
+                if img.mode != "RGB":
+                    img = img.convert("RGB")
+                cropped_img = img.crop(self._calculate_aspect_ratio_crop(img.size))  # Crop the image
+                cropped_img.save(io.BytesIO(), format="JPEG", quality=90)  # Save or manipulate as needed
+                # Flip the image vertically due to explorer client rendering images upside down
+                flipped_img = cropped_img.transpose(Image.FLIP_TOP_BOTTOM)
+                return np.array(flipped_img)
+        except Exception:
+            logger.exception("Error processing image")
+            raise
+
+    def _generate_deep_zoom_assets(self, image_array, folder_name):
+        """
+        Generate deep zoom assets from the image array.
+
+        Args:
+            image_array (np.ndarray): The image array.
+            folder_name (str): The name of the folder to save the assets.
+        """
+        h, w, bands = image_array.shape
+        linear = image_array.reshape(w * h * bands)
+        image = pyvips.Image.new_from_memory(linear.data, w, h, bands, "uchar")
+        image.dzsave(folder_name + "spatial", suffix=".jpeg")
```

Comment:
can we create a P1 performance ticket to see if we could use webp instead, so the tiles we store and download will be smaller 🙏 ?

Reply:
I've updated to save as WEBP

Reply:
Thanks so much! I just realized that we will also need to update the generated dzi file to specify Format "webp" instead of "jpeg"

For example, right now a dzi file looks like:

```dzi
<Image xmlns="http://schemas.microsoft.com/deepzoom/2008"
  Format="jpeg"  // <--- this needs to be webp
  Overlap="1"
  TileSize="254"
  >
  <Size 
    Height="1956"
    Width="1955"
  />
</Image>
````

Reply:
this is how it looks after the recent update:

```
<?xml version="1.0" encoding="UTF-8"?>
<Image xmlns="http://schemas.microsoft.com/deepzoom/2008"
  Format="webp"
  Overlap="1"
  TileSize="254"
  >
  <Size 
    Height="19720"
    Width="19720"
  />
</Image>
```

Reply:
Oh right, it generates the image files AND the manifest, so it makes sense that it'll update both 😂 Thanks for sharing!

Reply:
https://cellxgene.dev.single-cell.czi.technology/spatial-deep-zoom/0bb15784-1cea-47e1-9a00-57dcd127746c/spatial.dzi